x/mongo/driver/topology/pool.go

```diff
@@ -738,42 +762,61 @@ func (p *pool) removeConnection(conn *connection, reason reason, err error) erro
 // checkIn returns an idle connection to the pool. If the connection is perished or the pool is
 // closed, it is removed from the connection pool and closed.
 func (p *pool) checkIn(conn *connection) error {
-	if conn == nil {
-		return nil
-	}
-	if conn.pool != p {
-		return ErrWrongPool
-	}
+	return p.checkInWithCallback(conn, func() (reason, bool) {
+		if mustLogPoolMessage(p) {
+			keysAndValues := logger.KeyValues{
+				logger.KeyDriverConnectionID, conn.driverConnectionID,
+			}
 
-	if mustLogPoolMessage(p) {
-		keysAndValues := logger.KeyValues{
-			logger.KeyDriverConnectionID, conn.driverConnectionID,
+			logPoolMessage(p, logger.ConnectionCheckedIn, keysAndValues...)
 		}
 
-		logPoolMessage(p, logger.ConnectionCheckedIn, keysAndValues...)
-	}
-
-	if p.monitor != nil {
-		p.monitor.Event(&event.PoolEvent{
-			Type:         event.ConnectionCheckedIn,
-			ConnectionID: conn.driverConnectionID,
-			Address:      conn.addr.String(),
-		})
-	}
+		if p.monitor != nil {
+			p.monitor.Event(&event.PoolEvent{
+				Type:         event.ConnectionCheckedIn,
+				ConnectionID: conn.driverConnectionID,
+				Address:      conn.addr.String(),
+			})
+		}
 
-	return p.checkInNoEvent(conn)
+		r, perished := connectionPerished(conn)
+		if !perished && conn.pool.getState() == poolClosed {
+			perished = true
+			r = reason{
+				loggerConn: logger.ReasonConnClosedPoolClosed,
+				event:      event.ReasonPoolClosed,
+			}
+		}
+		return r, perished
+	})
 }
 
 // checkInNoEvent returns a connection to the pool. It behaves identically to checkIn except it does
 // not publish events. It is only intended for use by pool-internal functions.
 func (p *pool) checkInNoEvent(conn *connection) error {
+	return p.checkInWithCallback(conn, func() (reason, bool) {
+		r, perished := connectionPerished(conn)
+		if !perished && conn.pool.getState() == poolClosed {
+			perished = true
+			r = reason{
+				loggerConn: logger.ReasonConnClosedPoolClosed,
+				event:      event.ReasonPoolClosed,
+			}
+		}
+		return r, perished
+	})
+}
+
+func (p *pool) checkInWithCallback(conn *connection, cb func() (reason, bool)) error {
```

Comment:
It's not clear what the purpose of this callback is. The callback functions in `checkInNoEvent` and in `checkIn` are identical, except one logs and emits events and the other doesn't. Also, the callback param name doesn't say anything about what it's used for. We should refactor this to remove the duplicated code and either clarify or remove the callback param.

Reply:
I just realized Preston already addressed this. Resolving.

Reply:
`checkInWithCallback` is called at three places, each with some slight differences, like with or without logs/events and reasons to be returned. I think a callback is more flexible than multiple flags in the parameter list. However, I'll use a more illustrative name along with comments.